Sources/NIO/ByteBuffer-core.swift

```diff
@@ -29,6 +29,97 @@ let sysFree: @convention(c) (UnsafeMutableRawPointer?) -> Void = free
     }
 #endif
 
+@_versioned
+struct _UInt24: ExpressibleByIntegerLiteral {
+    typealias IntegerLiteralType = UInt16
+
+    @_versioned var b12: UInt16
+    @_versioned var b3: UInt8
+
+    private init(b12: UInt16, b3: UInt8) {
+        self.b12 = b12
+        self.b3 = b3
+    }
+
+    init(integerLiteral value: UInt16) {
+        self.init(b12: value, b3: 0)
+    }
+
+    static let bitWidth: Int = 24
+
+    static var max: _UInt24 {
+        return .init(b12: .max, b3: .max)
+    }
+
+    static let min: _UInt24 = 0
+}
+
+extension UInt32 {
+    init(_ value: _UInt24) {
+        var newValue: UInt32 = 0
+        newValue  = UInt32(value.b12)
+        newValue |= UInt32(value.b3) << 16
+        self = newValue
+    }
+}
+
+extension Int {
+    init(_ value: _UInt24) {
+        var newValue: Int = 0
+        newValue  = Int(value.b12)
+        newValue |= Int(value.b3) << 16
+        self = newValue
+    }
+}
+
+extension _UInt24 {
+    init(_ value: UInt32) {
+        precondition(value & 0xff_00_00_00 == 0, "value \(value) too large for _UInt24")
+        self.b12 = UInt16(truncatingIfNeeded: (value &    0xff_ff) >> 0 )
+        self.b3  =  UInt8(truncatingIfNeeded: (value & 0xff_ff_00) >> 16)
```

Comment:
You can save yourself a masking operation here. You already precondition that the top byte is zero, so with that knowledge in place you can just shift right by 16 bits and store into `UInt8`.

Reply:
Also as noted in the chat, the mask isn't really right anyway.

Reply:
oops! thanks